### DISCUSSION TITLE
Reusability and Next.js

### DISCUSSION THREAD

**Discussion by eighty4** [vercel/next.js#38620] [Help]
Is it possible to register api and component routes programmatically? Or to bundle frontend and backend code into reusable modules?

Examples include reusing ui/api code for comments, auth, etc.

**Comment by GabenGar** [comment]
What does "bundle frontend and backend code into reusable modules" even mean? If you meant writing isomorphic code the answer is yes and no. Yes because it's _technically_ possible, you just have to write your modules in a specific way. No, because _ALL_ dependencies of said modules have to be written this way, something both browser and node packages struggle with.
By the second sentence I assume you meant abstracting away auth logic in server functions. This can be accomplished by writing decorators for api/ssr/ssg functions, which accept a function with the signature you need and return a function with the signature nextjs needs.

**Reply by eighty4** [reply]
If there was a programmatic way to register API routes, then it wouldn't be necessary to create a bunch of files to map to the fns in your module. It doesn't get any more portable or modular to have the API routes just be a function. I was just wondering if there already exists or could exist an API to register those routes and bind functions to them without doing so with the director structure.

**Reply by GabenGar** [reply]
File-based routing is the whole point of NextJS. It's not about portability or modularity, it's about clear conventions and code-splitting.
Otherwise you can just do a catch-all route on the root of the `/api` folder and do whatever with it. This approach does mean you'll only have a single lambda function for API which grows in size as it gets more endpoints.

**Reply by eighty4** [reply]
Sure, but an important strength for any platform is extensibility. Next.js is incredible for what it achieves and how smoothly it does so. In the past week, however, closed extension points have caused hiccups on two use cases for the platform.
